README.md
# OpenFisca France — Fiscalité Minière

Système de fiscalité minière française modélisé avec OpenFisca

## Installation

Ce paquet requiert [Python 3.7](https://www.python.org/downloads/release/python-370/) et [pip](https://pip.pypa.io/en/stable/installing/).

Il est recommandé d'utiliser un [environnement virtuel](https://virtualenv.pypa.io/en/stable/) (_virtualenv_) avec un gestionnaire de _virtualenv_ tel que [Pew](https://github.com/berdario/pew).

### Installation pour le développement

Afin d'installer `OpenFisca-France-Fiscalite-Miniere`, il est nécessaire d'indexer son code et d'installer ses dépendances.

> Si vous disposez d'un environnement virtuel, activez-le à cette étape.

Afin d'installer l'ensemble des librairies nécessaires au développement, dans un terminal, exécuter le commande suivante :

```sh
make install
```

Pour en savoir plus sur cette commande, consulter le tag `install` et ses dépendances dans le fichier `./Makefile`.

### Installation pour la production

```sh
make install-prod
```

## Tester et vérifier une installation

Les tests du module `OpenFisca-France-Fiscalite-Miniere` peuvent être vérifiés par la commande suivante :

```sh
make test
```

Pour en savoir plus sur cette commande, consulter le tag `test` du fichier `./Makefile`.

Le résultat de cette commande doit afficher la ligne suivante :

```sh
...
Success: no issues found in X source files
...
============================== XX passed in X.XXs ==============================
```

🎉 Félicitations, `OpenFisca-France-Fiscalite-Miniere` est prêt à être utilisé !

## Servir l'API Web

Afin d'installer les librairies spécifiques à l'API Web, exécuter cette commande :

```sh
make install-api
```

Puis, démarrer l'API Web avec cette commande :

```sh
make serve
```

Un port par défaut est défini dans le `Makefile` (port `5000`).

Pour en savoir plus sur la commande `openfisca serve` sous-jascente et ses options, consultez la [documentation de référence](https://openfisca.org/doc/openfisca-python-api/openfisca_serve.html).

## Docker

Vous pouvez aussi utiliser Docker pour démarrer le serveur d'API Web. À l'intérieur du répertoire du dépôt `openfisca-france-fiscalite-miniere`, exécuter les commandes suivantes :
```sh
//Build de l’image localement
docker build --no-cache -t openfisca-france-fiscalite-miniere .

//Lancement de l’image
docker run --rm -p 5000:8000 openfisca-france-fiscalite-miniere
```
L'API Web est alors accessible via le port 5000 de l'hôte (8000 du conteneur).
Elle peut être testée dans un autre terminal. Elle renvoie par exemple un message d'accueil au format JSON en cas de requête `curl http://0.0.0.0:5000`.

### Envoyer une première requête à l'API Web

Un exemple de requête simple est donné dans ce fichier :
`./openfisca_france_fiscalite_miniere/examples/societe.json`

Pour transmettre la requête à l'API Web démarrée, dans un autre terminal, aller dans le répertoire du fichier JSON et envoyer la requête avec les commandes suivantes :

```
cd ./openfisca_france_fiscalite_miniere/examples
curl -X POST http://localhost:5000/calculate -H 'Content-Type: application/json' -d @societe.json
```

## Simulations

Le modèle `openfisca_france_fiscalite_miniere` est employé pour la simulation de réformes et l'estimation des effets de la loi sur la base de données de production minière protégées (non fournies).  
Les scripts de simulation sont assemblés dans le répertoire `./simulations`.  

### Réformes de la RCM

Il s'agit d'estimations des effets de réformes réalisées début 2020 sur la base de données [Camino](https://camino.beta.gouv.fr).

Ce sont des simulations d'une réforme de répartition communale de la Redevance Communale des Mines (RCM) pour les exploitations de chaque type de sel : pour chaque concession (titre) d'une mine, les simulations évaluent la distribution de la production et du produit de la taxe au prorata de la surface du titre sur chaque commune.

Elles sont rassemblées dans le répertoire `./simulations/reformes` :

* `test_essai_selg.py` est dédié au sel par abattage,
* `test_essai_selh.py` est dédié au sel en dissolution.

## Lexique

* RCM = Redevance Communale des Mines
* RDM = Redevance Départementale des Mines
* Substances M = Substances non énergétiques (métaux, minéraux)

## Références législatives complémentaires

* [Modalités de la déclaration de production minière à des fins fiscales depuis 1987](https://beta.legifrance.gouv.fr/codes/article_lc/LEGIARTI000006293414/1987-08-09)
* Unités prises en compte pour la RCM par produit extrait indiquées dans deux textes :
  - depuis 1987 : https://www.legifrance.gouv.fr/codes/article_lc/LEGIARTI000006293412/1987-08-09
  - depuis 1991 : https://www.legifrance.gouv.fr/codes/article_lc/LEGIARTI000006294877/2020-03-25


CHANGELOG.md
# Changelog

### 5.2.0 [#36](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/36)

* Ajoute les données pour la taxe de Guyane 2023
* Zones impactées :
  - `parameters/taxes/guyane/*`

### 5.1.0 [#31](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/31)

* Ajoute les redevances départementales et communales pour l'année 2023
* Zones impactées :
  - `parameters/redevances/*`
  
### 5.0.0 [#34](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/34)
* Change lithium_li20 par lithium pour garder une cohérence entre les paramètres et les variables
* Met à jour OpenFiscaCore et web en version 37.0.0
* Zones impactées :
  - `variables/redevances_communales_departementales/lithium.py`

### 4.3.1 [#32](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/32)
* Vérifie le formatage des fichiers python et yaml
* Zones impactées : '*.py && *.yml'


### 4.3.0 [#31](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/31)

* Ajoute les redevances départementales et communales pour l'année 2022
* Zones impactées :
  - `parameters/redevances/*`
* Détails :
  - Met à jour la version mineure d'OpenFisca (35.9.0)

### 4.2.0 [#29](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/29)

* Uniformisation des noms des variables des redevances communales et departementales
* Zones impactées :
  - `variables/redevances_communales_*/*.py`
  - `variables/frais.py`
  - `variables/redevances.py`
* Détails :
  - Enlève l'unité dans le nom de la classe
    - par exemple : `redevance_communale_des_mines_sel_raffine_kt` devient `redevance_communale_des_mines_sel_raffine`

### 4.1.0 [#30](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/30)

* Évolution du système socio-fiscal.
* Zones impactées :
  - `parameters/taxes/guyane/categories/*`
* Détails :
  - Revalorise les tarifs 2022 de la taxe sur l'or en Guyane.


# 4.0.0 [#27](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/27)

* Évolution du système socio-fiscal **non rétrocompatible**.
* Périodes concernées : à partir du 01/01/2020.
* Zones impactées :
  - `variables/redevances.py`
  - `variables/taxes.py`
* Détails :
  - Change `surface_totale` d'une caractéristique d'`Article` à une caractéristique de `Titre`
  - Ajoute le calcul des redevances (RDCM) à toutes les substances
    * Ajoute la `RCM` à toute substance dans `redevances_communales_departementales/`.
    * Ajoute la `RDM` à toute substance dans `redevances_communales_departementales/`.
  - Introduit `surface_communale_proportionnee` et l'introduit à partir de 2020 à toutes les substances.

### 3.0.2 [#26](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/26)

* Correction d'un crash.
* Périodes concernées : toutes.
* Zones impactées : `./variables/taxes.py`
* Détails :
  - Corrige l'erreur d'entité à l'appel de `taxe_guyane_brute` et `taxe_guyane_deduction`

### 3.0.1 [#25](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/25)

* Correction d'un crash.
* Périodes concernées : toutes.
* Zones impactées : N/A
* Détails :
  - Corrige l'erreur d'exécution de l'API Web (`cannot import name 'escape' from 'jinja2`)
  - Met à jour la dépendance `Flask` via la mise à jour d'`OpenFisca-Core`
  - Met à jour l'ignore des affichages console suite à l'évolution de `flake8-print`

# 3.0.0 [#22](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/22)

* Évolution du système socio-fiscal.
* Périodes concernées : toutes.
* Zones impactées :
  - `entities.py`.
  - `variables/*`
* Détails :
  - Renomme l'entité `Societe` en `Article`.
  - Ajoute l'entité `Titre`.
  - Ajoute des caractéristiques de titres.

### 2.0.2 [#21](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/21)

* Changement mineur.
* Périodes concernées : à partir du 01/01/2020.
* Zones impactées :
  - `simulations/drfip.py`
  - `simulations/estime_taxes_redevances.py`
* Détails :
  - Ajoute la génération des matrices 1403 et 1404 aux scripts de `simulations/`.
  - Matrices au format 2021. Code exécuté sur données de production 2020.

### 2.0.1 [#20](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/20)

* Changement mineur.
* Périodes concernées : à partir du 01/01/2020.
* Zones impactées :
  - `simulations/drfip.py`
  - `simulations/estime_taxes_redevances.py`
* Détails :
  - Adapte les scripts de production des matrices 1121 et 1122 au nouveau format des des données de production.
    * Adapte à l'évolution du format d'export des données de production de Camino.
    * Permet l'emploi des données de production 2019 au format 2020 et des données de production 2020 au format 2021.

# 2.0.0 [#19](https://github.com/openfisca/openfisca-france-fiscalite-miniere/pull/19)

* Évolution du système socio-fiscal.
* Périodes concernées : à partir du 01/01/2014.
* Zones impactées :
  - `parameters/taxes/guyane/categories/*`
  - `parameters/fiscalite/frais/*`
  - `variables/fiscalite.py` désormais `variables/frais.py`
* Détails :
  - Corrige les frais de gestions à partir de 2014.
    * le paramètre `fiscalite.frais.taux` est supprimé au bénéfice de `frais.taux_assiette_recouvrement` + `frais.taux_degrevement_non_valeur`.
  - Revalorise les tarifs 2021 de la taxe sur l'or en Guyane.

simulations/drfip.py
import numpy
import pandas  # noqa: I201

from simulations.sip import (
    get_sip_data,
    sip_guyane_cayenne, sip_guyane_cayenne_nom,
    sip_guyane_kourou, sip_guyane_kourou_nom,
    sip_guyane_st_laurent_du_maroni, sip_guyane_st_laurent_du_maroni_nom
    )


def build_designations_entreprises(data):
    nb_lignes = data.shape[0]
    tirets = numpy.full(nb_lignes, ' - ')
    texte_siren = numpy.full(nb_lignes, ' SIREN ')

    return (data["nom_entreprise"] + tirets
        + data["adresse_entreprise"] + tirets
        + data["siren"] + texte_siren)


def calculate_production_communale(data):
    # renseignements_orNet est renseignements_orExtrait à partir des data 2020
    return round(
        data["renseignements_orNet"] * (data["surface_communale"]
        / data["surface_totale"]),
        3)


def generate_matrice_drfip_guyane(data, annee_production, timestamp):
    '''
    Template MATRICE 1121
    REDEVANCE DÉPARTEMENTALE ET COMMUNALE DES MINES
    (articles 1519 et 1587 à 1589 du Code général des impôts)
    TAXE MINIÈRE SUR L'OR DE GUYANE
    (article 1599 quinquies B du Code général des impôts)
    '''
    colonnes_1121 = [
        "Numéro d'ordre de la matrice",  # [col. 1]
        "Commune du lieu principal d'exploitation",
        # [col. 2]
        # > Camino : commune de plus grande surface dans titres.communes
        # > au format commune_1 (float surface);commune_2 (float surface)
        "Désignation et adresse des concessionnaires,\
            titulaires de permis d’exploitation ou exploitants",
        # [col. 3]
        # (lorsqu’il s’agit d’une mine amodiée, porter à l’encre rouge la désignation
        # et l’adresse de l’exploitant au-dessous de celle du concessionnaire)
        # > Camino : adresse de l'amodiataire ou, à défaut, le titulaire.
        # > titres.amodiataires_noms/titres.amodiataires_adresses
        # > ou titres.titulaires_noms/titres.titulaires_adresses
        # > désignation en amodiataire/titulaire à mémoriser pour le type d'entreprise
        "Nature des substances extraites",
        # [col. 4]
        # > Camino : titres.domaine
        # Base des redevances :
        "Nature",  # [col. 5]
        # > Camino : titres.substances (séparateur ';')
        "Quantités (kg)",
        # [col. 6]
        # > Camino : OR activités.renseignements_orBrut (cellule potentiellement vide)
        # > renseignements_orExtrait à zéro ; ou renseignements_orNet
        #   (si brut non défini ?)
        # > pour activités.annee en N-1
        # > et activités.period se terminant par 'trimestre'
        #   (1er trimestre, 2e trimestre, 3e trimestre, 4e trimestre)
        # Redevance départementale :
        "Tarifs (RDM)",  # [col. 7]
        "Montant net (RDM)",  # [col. 8] (col. 6 x col. 7)
        # Redevance communale :
        "Tarifs (RCM)",  # [col. 9]
        "Montant net redevance des mines (RCM)",  # [col. 10] (col. 6 x col. 9)
        # ---
        "Total redevance des mines",  # [col. 11] (col.8 + col. 10)
        # Taxe minière sur l'or de Guyane :
        "PME",  # [col. 12] Tarifs par kg extrait pour les...
        "Autres entreprises",  # [col. 13] Tarifs par kg extrait pour les...
        # > Camino : titres.amodiataires_categorie ou titres.titulaires_categorie
        "Montant des investissements déduits",
        # [col. 14]
        # > Camino : somme des 4 rapports trimestriels année N-1
        #   (activités.renseignements_environnement)  # noqa: E800
        # > pour activités.annee en N-1 et activités.period en 'année'
        # > = sur l'interface web, activités,
        #   "Dépenses relatives à la protection de l’environnement (euros)"  # noqa: E800, E501
        "Montant net de taxe minière sur l'or de Guyane",
        # -> [col. 15] (col.6 x col.12)-col.14 ou (col.6 x col.13)-col.14
        # ---
        "Frais de gestion de la fiscalité directe locale",
        # -> [col. 16] (col.11 + col.15) x 8%
        "Service de la Direction générale des Finances publiques\
            en charge du recouvrement",
        # [col. 17]
        # > Camino : titre.administrations_noms ?
        "Numéro de l’article du rôle",  # [col. 18]
        "Observations, précisions"
        # [col. 19]
        # > Camino : activités.complement_texte ?
        ]  # une ligne par titre

    matrice = pandas.DataFrame(data, columns = colonnes_1121)

    numeros_ordre = data["titre_id"].index
    nb_lignes = len(numeros_ordre)

    matrice["Numéro d'ordre de la matrice"] = numeros_ordre + 1
    matrice["Commune du lieu principal d'exploitation"] = data[
        "commune_exploitation_principale"
        ]

    matrice[
        "Désignation et adresse des concessionnaires,\
            titulaires de permis d’exploitation ou exploitants"
        ] = build_designations_entreprises(data)

    matrice["Nature des substances extraites"] = numpy.full(
        nb_lignes, 'Minerais aurifères'
        )  # d'après texte législatif
    matrice["Nature"] = numpy.full(
        nb_lignes, "Kilogramme d'or contenu"
        )  # d'après texte législatif

    production_communale = calculate_production_communale(data)
    matrice["Quantités (kg)"] = (production_communale.astype(str))

    # Redevance départementale :
    matrice["Tarifs (RDM)"] = data["tarifs_rdm"]
    matrice["Montant net (RDM)"] = data[
        "redevance_departementale_des_mines_aurifere"
        ]

    # Redevance communale :
    matrice["Tarifs (RCM)"] = data["tarifs_rcm"]
    matrice["Montant net redevance des mines (RCM)"] = data[
        "redevance_communale_des_mines_aurifere"
        ]

    matrice["Total redevance des mines"] = round(
        matrice["Montant net (RDM)"] + matrice["Montant net redevance des mines (RCM)"],
        2)  # TODO vérifier pourquoi la somme a 4 décimales en l'absence d'arrondi

    # Taxe minière sur l'or de Guyane :
    matrice["PME"] = data["taxe_tarif_pme"]
    matrice["Autres entreprises"] = data["taxe_tarif_autres"]
    matrice["Montant des investissements déduits"] = data["investissement"]

    # TODO add check pour la taxe guyane pour l'ensemble des investissements déduits
    # PME et autres entreprises (non PME)
    matrice["Montant net de taxe minière sur l'or de Guyane"] = data["taxe_guyane"]

    matrice["Frais de gestion de la fiscalité directe locale"] = round(
        (matrice["Total redevance des mines"]
        + matrice["Montant net de taxe minière sur l'or de Guyane"]) * 0.08,
        2)

    matrice["Service de la Direction générale des Finances publiques en charge du recouvrement"] = data["drfip"]  # noqa: E501
    matrice["Numéro de l’article du rôle"] = data["titre_id"]
    matrice["Observations, précisions"] = data["observation"]

    matrice.to_csv(
        f'matrice_drfip_guyane_production_{annee_production}_{timestamp}.csv',
        index=False,
        sep=';',
        encoding='utf-8',
        decimal=','
        )


def generate_matrice_annexe_drfip_guyane(data, annee_production, timestamp):
    '''
    Template ÉTAT ANNEXE À LA MATRICE 1122
    REDEVANCE DÉPARTEMENTALE ET COMMUNALE DES MINES
    TAXE MINIÈRE SUR L'OR DE GUYANE
    ÉTAT ANNEXE À LA MATRICE
    DES  REDEVANCES, TAXES  ÉTABLIES  POUR  L'ANNÉE
    '''

    colonnes_1122 = [
        "Numéro d'ordre de la matrice",  # [col. 1]
        "Désignation des concessionnaires, exploitants ou explorateurs",  # [col. 2]
        "Désignation des concessions, permis ou explorations",  # [col. 3]
        # Départements et communes sur le territoire desquels
        # fonctionnent les exploitations :
        "Départements",  # [col. 4]
        "Communes",  # [col. 5]
        # Tonnages extraits au cours de l'année précédente :
        "par département",  # [col. 6]
        "par commune",  # [col. 7]
        "Observations"  # [col. 8]
        ]

    matrice_annexe = pandas.DataFrame(data, columns = colonnes_1122)

    numeros_ordre = data["titre_id"].index
    nb_lignes = len(numeros_ordre)

    matrice_annexe["Numéro d'ordre de la matrice"] = numeros_ordre + 1
    matrice_annexe[
        "Désignation des concessionnaires, exploitants ou explorateurs"
        ] = build_designations_entreprises(data)
    matrice_annexe["Désignation des concessions, permis ou explorations"] = data[
        "titre_id"
        ]
    # Départements et communes sur le territoire desquels
    # fonctionnent les exploitations :
    matrice_annexe["Départements"] = numpy.full(nb_lignes, 'Guyane')
    matrice_annexe["Communes"] = data['communes']
    # Tonnages extraits au cours de l'année précédente :
    matrice_annexe["par commune"] = calculate_production_communale(data)
    matrice_annexe["par département"] = matrice_annexe["par commune"]

    matrice_annexe["Observations"] = numpy.where(
        matrice_annexe["par commune"] > 0,
        "production en kilogramme d'or", "")

    matrice_annexe.to_csv(
        f'matrice_annexe_drfip_guyane_production_{annee_production}_{timestamp}.csv',
        index=False,
        sep=';',
        encoding='utf-8',
        decimal=','
        )


def generate_matrice_1403_drfip_guyane(data, annee_production, timestamp):
    colonnes_1403 = [
        "Service de la Direction générale des finances publiques en charge du recouvrement",  # Circonscription  # noqa: E501
        "Redevance départementale",  # Produit net = col. 2
        "Redevance communale",  # Produit net = col. 3
        "Taxe minière sur l'or de Guyane",  # Produit net = col. 4
        "Sommes revenant à Région de Guyane",
        "Sommes revenant à Conservatoire de biodiversité",  # Rien en 2020 ?

        # Sommes revenant à l'État :
        "Frais d'assiette et de recouvrement",  # = col. 7
        "Dégrèvements et non-valeurs",  # col. 8 / Rien en 2020 ?
        "Total des colonnes 7 et 8",  # col. 9
        "Total des colonnes 2, 3 ,4 et 9",
        "Nombre d'articles des rôles"
        ]

    matrice_1403 = pandas.DataFrame(columns = colonnes_1403)

    data_sip_cayenne = get_sip_data(
        data, "commune_exploitation_principale", sip_guyane_cayenne)
    data_sip_kourou = get_sip_data(
        data, "commune_exploitation_principale", sip_guyane_kourou)
    data_sip_st_laurent_du_maroni = get_sip_data(
        data, "commune_exploitation_principale", sip_guyane_st_laurent_du_maroni)

    matrice_1403[
        "Service de la Direction générale des finances publiques en charge du recouvrement"  # noqa: E501
        ] = [
        sip_guyane_cayenne_nom,
        sip_guyane_kourou_nom,
        sip_guyane_st_laurent_du_maroni_nom
        ]

    matrice_1403["Redevance départementale"] = [
        data_sip_cayenne[
            "redevance_departementale_des_mines_aurifere"
            ].sum().astype(int),
        data_sip_kourou[
            "redevance_departementale_des_mines_aurifere"
            ].sum().astype(int),
        data_sip_st_laurent_du_maroni[
            "redevance_departementale_des_mines_aurifere"
            ].sum().astype(int)
        ]

    matrice_1403["Redevance communale"] = [
        data_sip_cayenne["redevance_communale_des_mines_aurifere"].sum().astype(int),
        data_sip_kourou["redevance_communale_des_mines_aurifere"].sum().astype(int),
        data_sip_st_laurent_du_maroni[
            "redevance_communale_des_mines_aurifere"
            ].sum().astype(int)
        ]
    matrice_1403["Taxe minière sur l'or de Guyane"] = [
        data_sip_cayenne["taxe_guyane"].sum().astype(int),
        data_sip_kourou["taxe_guyane"].sum().astype(int),
        data_sip_st_laurent_du_maroni["taxe_guyane"].sum().astype(int)
        ]

    total_rdcm_taxe = (
        matrice_1403["Redevance départementale"]
        + matrice_1403["Redevance communale"]
        + matrice_1403["Taxe minière sur l'or de Guyane"]
        )

    matrice_1403["Sommes revenant à Région de Guyane"] = matrice_1403[
        "Taxe minière sur l'or de Guyane"
        ]
    matrice_1403["Sommes revenant à Conservatoire de biodiversité"] = [None, None, None]
    matrice_1403["Frais d'assiette et de recouvrement"] = (
        total_rdcm_taxe * 0.08
        ).astype(int)  # 4.4% mais à 8% en 2020, perte des centimes
    matrice_1403["Dégrèvements et non-valeurs"] = [None,
        None, None]  # 3.6% mais vide en 2020
    matrice_1403["Total des colonnes 7 et 8"] = matrice_1403[
        "Frais d'assiette et de recouvrement"
        ]
    matrice_1403["Total des colonnes 2, 3 ,4 et 9"] = total_rdcm_taxe + matrice_1403[
        "Total des colonnes 7 et 8"
        ]
    matrice_1403["Nombre d'articles des rôles"] = [
        data_sip_cayenne["nom_entreprise"].nunique(),
        data_sip_kourou["nom_entreprise"].nunique(),
        data_sip_st_laurent_du_maroni["nom_entreprise"].nunique()
        ]

    matrice_1403.to_csv(
        f'matrice_1403_drfip_guyane_production_{annee_production}_{timestamp}.csv',
        index=False,
        sep=';',
        encoding='utf-8',
        decimal=','
        )


def generate_matrice_1404_sip(columns_1404, data_sip, sip_name):
    matrice_1404_sip = pandas.DataFrame(columns = columns_1404)

    NB_ROWS_SIP = len(data_sip)

    matrice_1404_sip[
        "Service de la direction générale des finances publiques en charge du recouvrement"  # noqa: E501
        ] = [sip_name] * NB_ROWS_SIP
    matrice_1404_sip["Articles des rôles"] = data_sip["titre_id"].index

    # Avant affectation, on valide le bon alignement des articles
    # à défaut d'index commun aux DataFrame
    exploitants_sip_cayenne = build_designations_entreprises(data_sip)
    assert (matrice_1404_sip["Articles des rôles"].values
            == exploitants_sip_cayenne.index).all()
    matrice_1404_sip["Désignation des exploitants"] = exploitants_sip_cayenne.values

    # DEPARTEMENTS ET COMMUNES
    # sur les territoires desquels fonctionnent les exploitations :
    matrice_1404_sip["Départements"] = ['Guyane'] * NB_ROWS_SIP  # col. 4
    # col. 5
    matrice_1404_sip["Communes"] = data_sip["commune_exploitation_principale"].values

    # ELEMENTS SERVANT DE BASE A LA REPARTITION pour chaque exploitation :
    matrice_1404_sip["Revenus imposables à la TFPB (état 1123, col. 3)"] = [
        None] * NB_ROWS_SIP  # vide en drfip

    # Avant affectation, on valide le bon alignement des articles
    # à défaut d'index commun aux DataFrame
    production_communale = calculate_production_communale(data_sip)
    assert (matrice_1404_sip["Articles des rôles"].values
            == production_communale.index).all()
    matrice_1404_sip["Tonnages extraits :"] = production_communale.values

    # REDEVANCE DÉPARTEMENTALE :
    matrice_1404_sip["Produit net de la redevance (RDM)"] = data_sip[
        "redevance_departementale_des_mines_aurifere"
        ].values
    matrice_1404_sip[
        "Sommes revenant aux départements désignés dans la colonne 4 (a)"
        ] = matrice_1404_sip["Produit net de la redevance (RDM)"]

    # REDEVANCE COMMUNALE :
    # col. 10
    matrice_1404_sip["Produit net de la redevance (RCM)"] = data_sip[
        "redevance_communale_des_mines_aurifere"
        ].values
    # > Répartition
    matrice_1404_sip["1ère fraction (col. 10 x 35%)"] = matrice_1404_sip[
        "Produit net de la redevance (RCM)"
        ] * 0.35
    matrice_1404_sip["2ème fraction (col. 10 x 10%)"] = matrice_1404_sip[
        "Produit net de la redevance (RCM)"
        ] * 0.1
    matrice_1404_sip["3ème fraction (col. 10 x 55%)"] = matrice_1404_sip[
        "Produit net de la redevance (RCM)"
        ] * 0.55
    # > Sommes revenant aux communes désignées dans la colonne 5
    # au titre des 1ère et 2ème fractions
    # col. 14
    matrice_1404_sip["1ère fraction (b)"] = matrice_1404_sip[
        "1ère fraction (col. 10 x 35%)"
        ]
    # col. 15
    matrice_1404_sip["2ème fraction (a)"] = matrice_1404_sip[
        "2ème fraction (col. 10 x 10%)"
        ]
    matrice_1404_sip["Total (col. 14 + col. 15)"] = matrice_1404_sip[
        "1ère fraction (b)"
        ] + matrice_1404_sip["2ème fraction (a)"]

    # TAXE MINIÈRE SUR L'OR DE GUYANE
    matrice_1404_sip["Produit net de la taxe"] = data_sip["taxe_guyane"].values
    # > Répartition
    matrice_1404_sip["Région de Guyane"] = matrice_1404_sip["Produit net de la taxe"]
    matrice_1404_sip["Conservatoire"] = [None] * NB_ROWS_SIP  # vide en 2020
    matrice_1404_sip["Observations"] = [None] * NB_ROWS_SIP

    return matrice_1404_sip


def generate_matrices_1404_drfip_guyane(data, annee_production, timestamp):
    colonnes_1404 = [
        "Service de la direction générale des finances publiques en charge du recouvrement",  # SIP  # noqa: E501
        "Articles des rôles",
        "Désignation des exploitants",

        # DEPARTEMENTS ET COMMUNES
        # sur les territoires desquels fonctionnent les exploitations :
        "Départements",  # col. 4
        "Communes",  # col. 5

        # ELEMENTS SERVANT DE BASE A LA REPARTITION pour chaque exploitation :
        "Revenus imposables à la TFPB (état 1123, col. 3)",  # col. 6 / vide en drfip
        "Tonnages extraits :",  # col. 7

        # REDEVANCE DÉPARTEMENTALE :
        "Produit net de la redevance (RDM)",
        "Sommes revenant aux départements désignés dans la colonne 4 (a)",

        # REDEVANCE COMMUNALE :
        "Produit net de la redevance (RCM)",  # col. 10
        # > Répartition
        "1ère fraction (col. 10 x 35%)",
        "2ème fraction (col. 10 x 10%)",
        "3ème fraction (col. 10 x 55%)",
        # > Sommes revenant aux communes désignées dans la colonne 5
        # au titre des 1ère et 2ème fractions
        "1ère fraction (b)",  # col. 14
        "2ème fraction (a)",  # col. 15
        "Total (col. 14 + col. 15)",

        # TAXE MINIÈRE SUR L'OR DE GUYANE
        "Produit net de la taxe",
        # > Répartition
        "Région de Guyane",
        "Conservatoire",  # vide en 2020
        "Observations"
        ]
    # où :
    # (a) Attributions faites au prorata des tonnages indiqués dans la colonne 7.
    # (b) Attributions faites au prorata des revenus indiqués dans la colonne 6.
    # (c) Le produit net de la taxe sur l'or de Guyane est affecté en totalité
    # à la région de Guyane en l'absence de création du conservatoire
    # (d) Dès que la création du conservatoire sera intervenu le produit de la taxe
    # due par les PME sera réparti à parts égales entre la région Guyane
    # et cet organisme et la taxe due par les autres

    data_sip_cayenne = get_sip_data(
        data, "commune_exploitation_principale", sip_guyane_cayenne)
    matrice_1404_sip_cayenne = generate_matrice_1404_sip(
        colonnes_1404, data_sip_cayenne, sip_guyane_cayenne_nom)
    matrice_1404_sip_cayenne.to_csv(
        f'matrice_1404_sip_guyane_cayenne_production_{annee_production}_{timestamp}.csv',  # noqa: E501
        index=False,
        sep=';',
        encoding='utf-8',
        decimal=','
        )

    data_sip_kourou = get_sip_data(
        data, "commune_exploitation_principale", sip_guyane_kourou)
    matrice_1404_sip_kourou = generate_matrice_1404_sip(
        colonnes_1404, data_sip_kourou, sip_guyane_kourou_nom)
    matrice_1404_sip_kourou.to_csv(
        f'matrice_1404_sip_guyane_kourou_production_{annee_production}_{timestamp}.csv',
        index=False,
        sep=';',
        encoding='utf-8',
        decimal=','
        )

    data_sip_st_laurent_du_maroni = get_sip_data(
        data, "commune_exploitation_principale", sip_guyane_st_laurent_du_maroni
        )
    matrice_1404_sip_st_laurent_du_maroni = generate_matrice_1404_sip(
        colonnes_1404, data_sip_st_laurent_du_maroni,
        sip_guyane_st_laurent_du_maroni_nom
        )
    matrice_1404_sip_st_laurent_du_maroni.to_csv(
        f'matrice_1404_sip_guyane_st_laurent_du_maroni_production_{annee_production}_{timestamp}.csv',  # noqa: E501
        index=False,
        sep=';',
        encoding='utf-8',
        decimal=','
        )
